### PR TITLE
[Malleability Immutable] Removed non-constructor mutations for Chunk, ChunkDataPack and ExecutionResult structs.

### PR DESCRIPTION
**Pull request by AndriiDiachuk**
closes: https://github.com/onflow/flow-go/issues/7279 https://github.com/onflow/flow-go/issues/7280 https://github.com/onflow/flow-go/issues/7290

## Context
In this PR were added constructors with validity checks for `Chunk`, `ChunkDataPack` and `ExecutionResult` structs, removed non-constructor mutations for those structs.

### CODE REVIEW

File: engine/common/rpc/convert/execution_results.go [hunk 1]
```diff
@@ -183,20 +183,52 @@ func MessageToChunk(m *entities.Chunk) (*flow.Chunk, error) {
...
-	return &flow.Chunk{
-		ChunkBody: chunkBody,
-		Index:     m.Index,
-		EndState:  endState,
-	}, nil
+
+	serviceEventCountPtr := MessageToServiceEventCountField(m.ServiceEventCount)
+	var chunk *flow.Chunk
+
+	// Branch on nil-vs-non-nil to preserve backward compatibility
```
Comment (jordanschalm) [inline engine/common/rpc/convert/execution_results.go:196]:
```suggestion
	// Branch on nil-vs-non-nil to preserve backward compatibility
	//TODO(mainnet27, #6773): remove the v1 branch here
```

File: engine/execution/block_result.go [hunk 2]
```diff
@@ -191,17 +192,22 @@ func (ar *BlockAttestationResult) AppendCollectionAttestationResult(
...
+func (ar *BlockAttestationResult) AllChunks() ([]*flow.Chunk, error) {
 	chunks := make([]*flow.Chunk, len(ar.collectionAttestationResults))
 	for i := 0; i < len(ar.collectionAttestationResults); i++ {
-		chunks[i] = ar.ChunkAt(i) // TODO(ramtin): cache and optimize this
+		chunk, err := ar.ChunkAt(i)
+		if err != nil {
+			return nil, fmt.Errorf("could not find chunk: %w", err)
+		}
+
+		chunks[i] = chunk // TODO(ramtin): cache and optimize this
```
Comment (jordanschalm) [inline engine/execution/block_result.go]:
```suggestion
		chunks[i] = chunk // TODO(ramtin): cache and optimize this
```

File: engine/execution/block_result.go [hunk 3]
```diff
@@ -214,29 +220,43 @@ func (ar *BlockAttestationResult) ChunkAt(index int) *flow.Chunk {
...
+			return nil, fmt.Errorf("could not find chunk data pack: %w", err)
+		}
+		chunkDataPacks[i] = chunkDataPack // TODO(ramtin): cache and optimize this
 	}
-	return chunkDataPacks
+	return chunkDataPacks, nil
 }
 
-func (ar *BlockAttestationResult) ChunkDataPackAt(index int) *flow.ChunkDataPack {
+func (ar *BlockAttestationResult) ChunkDataPackAt(index int) (*flow.ChunkDataPack, error) {
```
Comment (jordanschalm) [inline engine/execution/block_result.go:260]:
```suggestion
// ChunkDataPackAt returns the ChunkDataPack for the collection at the given index.
// Receiver BlockAttestationResult is expected to be well-formed; callers must use an index that exists.
// No errors are expected during normal operation. 
func (ar *BlockAttestationResult) ChunkDataPackAt(index int) (*flow.ChunkDataPack, error) {
```

File: engine/execution/block_result.go [hunk 4]
```diff
@@ -191,17 +192,22 @@ func (ar *BlockAttestationResult) AppendCollectionAttestationResult(
...
+		}
+
+		chunks[i] = chunk // TODO(ramtin): cache and optimize this
 	}
-	return chunks
+	return chunks, nil
 }
 
-func (ar *BlockAttestationResult) ChunkAt(index int) *flow.Chunk {
+func (ar *BlockAttestationResult) ChunkAt(index int) (*flow.Chunk, error) {
```
Comment (jordanschalm) [inline engine/execution/block_result.go:210]:
```suggestion
// ChunkAt returns the Chunk for the collection at the given index.
// Receiver BlockAttestationResult is expected to be well-formed; callers must use an index that exists.
// No errors are expected during normal operation. 
func (ar *BlockAttestationResult) ChunkAt(index int) (*flow.Chunk, error) {
```

File: model/chunks/chunks.go [hunk 5]
```diff
@@ -4,14 +4,11 @@ import (
 	"github.com/onflow/flow-go/model/flow"
 )
 
-// ChunkListFromCommit creates a chunklist with one chunk whos final state is
+// ChunkListFromCommit creates a chunklist with one chunk who's final state is
```
Comment (jordanschalm) [inline model/chunks/chunks.go]:
```suggestion
// ChunkListFromCommit creates a chunklist with one chunk whose final state is
```
😅

File: model/flow/chunk.go [hunk 6]
```diff
@@ -160,69 +172,103 @@ func (ch *Chunk) EncodeRLP(w io.Writer) error {
...
+			ServiceEventCount:    untrusted.ServiceEventCount,
+			TotalComputationUsed: untrusted.TotalComputationUsed,
+		},
+		Index:    untrusted.Index,
+		EndState: untrusted.EndState,
+	}, nil
+}
+
+// NewCommitChunk creates a cnunk who's final state is the commit
+func NewCommitChunk(
```
Comment (jordanschalm) [inline model/flow/chunk.go]:
```suggestion
// NewRootChunk creates a chunk whose final state is the given commit, with all other fields set to zero.
// This is a special kind of chunk used only as the sole chunk of a root execution result, which forms
// a part of the root protocol state snapshot used as the trusted root for a spork. 
func NewRootChunk(
```

File: model/flow/chunk_test.go [hunk 7]
```diff
@@ -60,57 +60,75 @@ func TestChunkList_Indices(t *testing.T) {
...
-	)
 
+	var serviceCount uint16 = 0
+
+	chunk, err := flow.NewChunk(flow.UntrustedChunk{
+		ChunkBody: flow.ChunkBody{
+			CollectionIndex:      3,
+			StartState:           unittest.StateCommitmentFixture(),
+			EventCollection:      unittest.IdentifierFixture(),
+			ServiceEventCount:    &serviceCount,
```
Comment (jordanschalm) [inline model/flow/chunk_test.go]:
```suggestion
	chunk, err := flow.NewChunk(flow.UntrustedChunk{
		ChunkBody: flow.ChunkBody{
			CollectionIndex:      3,
			StartState:           unittest.StateCommitmentFixture(),
			EventCollection:      unittest.IdentifierFixture(),
			ServiceEventCount:    unittest.PtrTo[uint16](0),
```
Can use this helper function to take the reference

File: model/flow/execution_result.go [hunk 8]
```diff
@@ -17,20 +18,45 @@ type ExecutionResult struct {
...
+func NewExecutionResult(untrusted UntrustedExecutionResult) (*ExecutionResult, error) {
+	if untrusted.PreviousResultID == ZeroID {
+		return nil, fmt.Errorf("PreviousResultID must not be empty")
+	}
+
+	if untrusted.BlockID == ZeroID {
+		return nil, fmt.Errorf("BlockID must not be empty")
 	}
+
+	if untrusted.Chunks == nil {
```
Comment (jordanschalm) [inline model/flow/execution_result.go]:
```suggestion
	if len(untrusted.Chunks) == 0 {
```

File: model/flow/execution_result.go [hunk 9]
```diff
@@ -17,20 +18,45 @@ type ExecutionResult struct {
 	ExecutionDataID  Identifier // hash commitment to flow.BlockExecutionDataRoot
```
Comment (jordanschalm) [inline model/flow/execution_result.go:20]: Missing the `//structwrite:immutable - mutations allowed only within the constructor` annotation for ExecutionResult

File: cmd/bootstrap/run/result.go [hunk 10]
```diff
@@ -11,7 +11,7 @@ func GenerateRootResult(
 	epochSetup *flow.EpochSetup,
 	epochCommit *flow.EpochCommit,
 ) *flow.ExecutionResult {
-
+	//nolint:structwrite
```
Comment (AndriiDiachuk) [inline cmd/bootstrap/run/result.go]: @jordanschalm Can we create here another constructor like `NewRootExecutionResult`, which would skip validity checks for `PreviousResultID` and `ExecutionDataID` fields ?

File: engine/execution/block_result.go [hunk 11]
```diff
@@ -214,29 +222,46 @@ func (ar *BlockAttestationResult) ChunkAt(index int) *flow.Chunk {
...
+			CollectionIndex:      uint(index),
+			StartState:           attestRes.startStateCommit,
+			EventCollection:      attestRes.eventCommit,
+			ServiceEventCount:    serviceEventCountPtr,
+			TotalComputationUsed: execRes.executionSnapshot.TotalComputationUsed(),
+			NumberOfTransactions: uint64(len(execRes.TransactionResults())),
+		},
+		Index:    uint64(index),
+		EndState: attestRes.endStateCommit,
+	})
```
Comment (durkmurder) [inline engine/execution/block_result.go]:
```suggestion
	chunk, err := ar.versionAwareChunkConstructor(flow.UntrustedChunk{
		ChunkBody: flow.ChunkBody{
			BlockID:              ar.Block.ID(),
			CollectionIndex:      uint(index),
			StartState:           attestRes.startStateCommit,
			EventCollection:      attestRes.eventCommit,
			ServiceEventCount: convert.MessageToServiceEventCountField(uint32(ar.ServiceEventCountForChunk(index))),
			TotalComputationUsed: execRes.executionSnapshot.TotalComputationUsed(),
			NumberOfTransactions: uint64(len(execRes.TransactionResults())),
		},
		Index:    uint64(index),
		EndState: attestRes.endStateCommit,
	})
```

Comment (durkmurder) [reply engine/execution/block_result.go]: This also fixes formatting of TODO statement above.

File: cmd/bootstrap/run/result.go (hunk 10 above)
Comment (jordanschalm) [reply cmd/bootstrap/run/result.go]: Good idea - yes let's do that.

File: model/flow/execution_result_test.go [hunk 12]
```diff
@@ -190,3 +190,166 @@ func TestExecutionResult_ServiceEventsByChunk(t *testing.T) {
 		}
 	})
 }
+
+// TestNewExecutionResult verifies that NewExecutionResult constructs a valid
+// ExecutionResult when given all required fields, and returns an error if any
+// required field is missing.
+// This test covers the test cases:
```
Comment (durkmurder) [inline model/flow/execution_result_test.go]: I don't think this description adds clarity, it just names of the sub-tests without extra info. I would suggest to expand each of them to give little context like [here](https://github.com/onflow/flow-go/pull/7527/files#diff-c5779bd215a1acc2b9d1c0bdf56e9c2e5e48e3e065b7cb593636003707ee121aR99).

File: model/flow/execution_result_test.go [hunk 13]
```diff
@@ -190,3 +190,166 @@ func TestExecutionResult_ServiceEventsByChunk(t *testing.T) {
...
+			Chunks:           chunks,
+			ExecutionDataID:  validExecDataID,
+		}
+		res, err := flow.NewExecutionResult(u)
+		assert.Error(t, err)
+		assert.Nil(t, res)
+		assert.Contains(t, err.Error(), "BlockID")
+	})
+
+	t.Run("nil Chunks", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/execution_result_test.go:282]: also need to test if chunks is empty

File: model/flow/execution_result_test.go [hunk 14]
```diff
@@ -190,3 +190,166 @@ func TestExecutionResult_ServiceEventsByChunk(t *testing.T) {
...
+			ExecutionDataID:  validExecDataID,
+		}
+		res, err := flow.NewExecutionResult(u)
+		assert.NoError(t, err)
+		assert.NotNil(t, res)
+		assert.Equal(t, validPrevID, res.PreviousResultID)
+		assert.Equal(t, validBlockID, res.BlockID)
+		assert.Equal(t, chunks, res.Chunks)
+		assert.Equal(t, flow.ServiceEventList{}, res.ServiceEvents)
+		assert.Equal(t, validExecDataID, res.ExecutionDataID)
```
Comment (durkmurder) [inline model/flow/execution_result_test.go]:
```suggestion
		assert.Equal(t, *res, flow.ExecutionResult(u))
```
since they are of the same type you can do something like this.

File: model/flow/execution_result_test.go [hunk 15]
```diff
@@ -190,3 +190,166 @@ func TestExecutionResult_ServiceEventsByChunk(t *testing.T) {
...
+		assert.Error(t, err)
+		assert.Nil(t, res)
+		assert.Contains(t, err.Error(), "ExecutionDataID")
+	})
+}
+
+// TestNewRootExecutionResult verifies that NewRootExecutionResult constructs a valid root
+// ExecutionResult when given all required fields, and returns an error if any
+// required field is missing.
+// This test covers the test cases:
```
Comment (durkmurder) [inline model/flow/execution_result_test.go]: same here

File: model/flow/execution_result_test.go [hunk 16]
```diff
@@ -190,3 +190,166 @@ func TestExecutionResult_ServiceEventsByChunk(t *testing.T) {
...
+			PreviousResultID: flow.ZeroID,
+			BlockID:          validBlockID,
+			Chunks:           chunks,
+			ServiceEvents:    flow.ServiceEventList{},
+			ExecutionDataID:  flow.ZeroID,
+		}
+		res, err := flow.NewRootExecutionResult(u)
+		assert.NoError(t, err)
+		assert.NotNil(t, res)
+		assert.Equal(t, flow.ZeroID, res.PreviousResultID)
```
Comment (durkmurder) [inline model/flow/execution_result_test.go]: same here

File: model/flow/execution_result_test.go [hunk 17]
```diff
@@ -190,3 +190,166 @@ func TestExecutionResult_ServiceEventsByChunk(t *testing.T) {
...
+			ServiceEvents:    flow.ServiceEventList{},
+			ExecutionDataID:  validExecDataID,
+		}
+		res, err := flow.NewRootExecutionResult(u)
+		assert.Error(t, err)
+		assert.Nil(t, res)
+		assert.Contains(t, err.Error(), "BlockID")
+	})
+
+	t.Run("nil Chunks", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/execution_result_test.go:375]: add a test for an empty chunks.

File: model/flow/chunk_test.go [hunk 18]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
 		}),
 	)
 }
+
+// TestNewChunkDataPack verifies that NewChunkDataPack constructs a valid ChunkDataPack
+// from complete, non-zero inputs and returns errors when any required field is missing.
+// This test covers the test cases:
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: same here

File: model/flow/chunk_test.go [hunk 19]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
...
+		untrusted := baseChunkDataPack
+		untrusted.StartState = flow.StateCommitment{}
+
+		pack, err := flow.NewChunkDataPack(untrusted)
+		assert.Error(t, err)
+		assert.Nil(t, pack)
+		assert.Contains(t, err.Error(), "StartState")
+	})
+
+	t.Run("empty Proof", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: this is nil proof, add a test for empty as well.

File: model/flow/chunk_test.go [hunk 20]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
...
+		untrusted := baseChunkDataPack
+		untrusted.ExecutionDataRoot.BlockID = flow.ZeroID
+
+		pack, err := flow.NewChunkDataPack(untrusted)
+		assert.Error(t, err)
+		assert.Nil(t, pack)
+		assert.Contains(t, err.Error(), "ExecutionDataRoot.BlockID")
+	})
+
+	t.Run("empty ExecutionDataRoot.ChunkExecutionDataIDs", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: also test if empty.

File: model/flow/chunk_test.go [hunk 21]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
...
+		assert.Error(t, err)
+		assert.Nil(t, pack)
+		assert.Contains(t, err.Error(), "ExecutionDataRoot.ChunkExecutionDataIDs")
+	})
+}
+
+// TestNewChunk verifies that NewChunk constructs a valid Chunk when given
+// complete, nonzero fields, and returns an error if any required field is
+// missing or zero.
+// It covers:
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: same here

File: model/flow/chunk_test.go [hunk 22]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
...
+		},
+		Index:    1,
+		EndState: validState,
+	}
+
+	t.Run("valid chunk", func(t *testing.T) {
+		ch, err := flow.NewChunk(base)
+		assert.NoError(t, err)
+		assert.NotNil(t, ch)
+		assert.Equal(t, validID, ch.ChunkBody.BlockID)
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: same here, you can simplify this boilerplate a lot

File: model/flow/chunk_test.go [hunk 23]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
...
+		assert.Error(t, err)
+		assert.Nil(t, ch)
+		assert.Contains(t, err.Error(), "EndState")
+	})
+}
+
+// TestNewChunk_ProtocolVersion1 verifies that NewChunk_ProtocolVersion1 constructs a
+// valid Chunk for protocol version 1 when given complete, non-zero fields, and
+// returns an error if any required field is missing.
+// This test covers the test cases:
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: same here

File: model/flow/chunk_test.go [hunk 24]
```diff
@@ -376,3 +388,275 @@ func TestChunkDataPackMalleability(t *testing.T) {
...
+	t.Run("valid protocol v1 chunk", func(t *testing.T) {
+		ch, err := flow.NewChunk_ProtocolVersion1(base)
+		assert.NoError(t, err)
+		assert.NotNil(t, ch)
+
+		// ServiceEventCount must be nil for protocol v1
+		assert.Nil(t, ch.ChunkBody.ServiceEventCount)
+
+		// Other fields should match
+		assert.Equal(t, validID, ch.ChunkBody.BlockID)
```
Comment (durkmurder) [inline model/flow/chunk_test.go]: can be simplified.
